ksqldb-engine/src/main/java/io/confluent/ksql/engine/RuntimeAssignor.java

```diff
@@ -0,0 +1,124 @@
+/*
+ * Copyright 2021 Confluent Inc.
+ *
+ * Licensed under the Confluent Community License (the "License"); you may not use
+ * this file except in compliance with the License.  You may obtain a copy of the
+ * License at
+ *
+ * http://www.confluent.io/confluent-community-license
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OF ANY KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package io.confluent.ksql.engine;
+
+import static io.confluent.ksql.util.QueryApplicationId.buildSharedRuntimeId;
+
+import io.confluent.ksql.name.SourceName;
+import io.confluent.ksql.query.QueryId;
+import io.confluent.ksql.util.BinPackedPersistentQueryMetadataImpl;
+import io.confluent.ksql.util.KsqlConfig;
+import io.confluent.ksql.util.PersistentQueryMetadata;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+public class RuntimeAssignor {
+  private final Map<String, Collection<SourceName>> runtimesToSources;
+  private final Map<QueryId, String> idToRuntime;
+
+  public RuntimeAssignor(final KsqlConfig config) {
+    this(config, 8);
+  }
+
+  public RuntimeAssignor(final KsqlConfig config, final int runtimes) {
+    runtimesToSources = new HashMap<>();
+    idToRuntime = new HashMap<>();
+    for (int i = 0; i < runtimes; i++) {
+      final String runtime = buildSharedRuntimeId(config, true, i);
+      runtimesToSources.put(runtime, new HashSet<>());
+    }
+  }
+
+  private RuntimeAssignor(final RuntimeAssignor other) {
+    this.runtimesToSources = new HashMap<>();
+    this.idToRuntime = new HashMap<>();
+    for (Map.Entry<String, Collection<SourceName>> runtime
+        : other.runtimesToSources.entrySet()) {
+      this.runtimesToSources.put(runtime.getKey(), new HashSet<>(runtime.getValue()));
+    }
+    this.idToRuntime.putAll(other.idToRuntime);
+  }
+
+  public RuntimeAssignor createSandbox() {
+    return new RuntimeAssignor(this);
+  }
+
+  public String getRuntime(final QueryId queryId,
```

Comment:
`getX` often implies that the method is read-only. Consider using a name like `getAndMaybeAddRuntimeForQuery`

Reply:
Yes, that is a better name